### PR TITLE
Optimize memory usage in samples/repositories import [SCI-1665]

### PR DESCRIPTION
**Pull request by okriuchykhin**
https://biosistemika.atlassian.net/browse/SCI-1665

### CODE REVIEW

File: app/models/repository.rb [hunk 1]
```diff
@@ -130,54 +131,69 @@ def import_records(sheet, mappings, user)
     unless col_compact.map(&:id).uniq.length == col_compact.length
       return { status: :error, nr_of_added: nr_of_added, total_nr: total_nr }
     end
+    rows = if sheet.is_a?(Roo::CSV)
```
Comment (ZmagoD) [inline app/models/repository.rb]: you can do better than this :smile:

File: app/models/repository.rb [hunk 2]
```diff
@@ -130,54 +131,69 @@ def import_records(sheet, mappings, user)
...
-        record_row = RepositoryRow.new(name: sheet.row(i)[name_index],
-                                   repository: self,
-                                   created_by: user,
-                                   last_modified_by: user)
-        record_row.transaction(requires_new: true) do
-          unless record_row.save
-            errors = true
-            raise ActiveRecord::Rollback
-          end
+    rows.each do |row|
```
[the remark this answers is not in the document]
Comment (ZmagoD) [reply app/models/repository.rb]: find_each works here?

File: app/services/import_repository/parse_repository.rb [hunk 3]
```diff
@@ -10,43 +10,57 @@ def initialize(options)
 
     def data
       # Get data (it will trigger any errors as well)
-      header = @sheet.row(1)
-      columns = @sheet.row(2)
+      if @sheet.is_a?(Roo::CSV)
```
Comment (ZmagoD) [inline app/services/import_repository/parse_repository.rb]: I seen this code multiple time can you extract this in one method

File: Gemfile.lock [hunk 4]
```diff
@@ -362,6 +365,7 @@ DEPENDENCIES
   bootstrap_form
   byebug
   commit_param_routing
+  creek
```
[the remark this answers is not in the document]
Comment (okriuchykhin) [reply Gemfile.lock]: XLSX parser

File: app/controllers/repositories_controller.rb [hunk 5]
```diff
@@ -198,12 +198,15 @@ def parse_sheet
...
         repository_response(t('general.file.size_exceeded',
                               file_size: Constants::FILE_MAX_SIZE_MB))
-      elsif parsed_file.empty?
-        flash[:notice] = t('teams.parse_sheet.errors.empty_file')
-        redirect_to back and return
       else
         @import_data = parsed_file.data
-        if parsed_file.generated_temp_file?
+
+        unless @import_data.header.any? && @import_data.columns.any?
```
[the remark this answers is not in the document]
Comment (okriuchykhin) [reply app/controllers/repositories_controller.rb]: For the array?)

File: app/models/repository.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (okriuchykhin) [reply app/models/repository.rb]: No, this is not an ActiveRecord, just an enumerator

File: app/controllers/repositories_controller.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (okriuchykhin) [reply app/controllers/repositories_controller.rb]: these are just arrays

[the remark this answers is not in the document]
Comment (ZmagoD) [reply app/controllers/repositories_controller.rb]: sorry use empty?

[the remark this answers is not in the document]
Comment (okriuchykhin) [reply app/controllers/repositories_controller.rb]: Isn't 'if empty?' the same as 'unless any?'

**Pull request merged**